### PR TITLE
Merge REST API server branch to master

### PR DESCRIPTION
**Pull request by mgolosova**
This PR originally was meant to introduce some basic functionality of the API server, but now this branch (`api`) is considered to be an independent branch, where th server application is being developed.

It was never merged to `master` for two reasons:
* the development was never finished -- there's always something to do;
* there was no need to merge (no functionality from `api` is needed in `master`).

But these two branches share one thing: ES mapping. Some changes were made to it in the `api` and never get to `master` -- and it was never a problem... until today. Now I need to make some changes to the mapping in `master` that requires usage of our custom analyzer, which is `task_dataset_anamyzer` in `master` and `dsname` in `api`).

So I suggest to merge this PR now, and open a new one whenever `api` changes mapping:
* now -- to avoid conflicts in analyzers in the future (for sooner or later the mapping will be merged);
* now and later -- to have the most actual version of mapping in `master` [1].

[1] For it is more natural to take it from here when we need to load new version to ES instance, and I personally was quite surprised to realize that the mapping currently used by ES instance is not the one provided by `master`...

---

**Old description**

In this PR only the basic functionality of the server is to be added:
 - [x] skeleton of the functions and modules layout (`NotImplemented` placeholders are good enough)
 - [x] query string parsing
 - [x] error handling
 - [ ] ~default method (server info)~ *(one day it will happen, but not now)*

### CODE REVIEW

**Comment by mgolosova** [comment]
[Установить API-сервер на aiatlas172](https://trello.com/c/Zjx7t3o2/37-%D1%83%D1%81%D1%82%D0%B0%D0%BD%D0%BE%D0%B2%D0%B8%D1%82%D1%8C-api-%D1%81%D0%B5%D1%80%D0%B2%D0%B5%D1%80-%D0%BD%D0%B0-aiatlas172)

File: Utils/API/server/lib/dkb/api/exceptions.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+class NotFoundException(DkbApiException):
+    """ Base exception for "smth not found" errors. """
+    code = 400
+    details = "For list of available methods and (sub)categories:" \
+              " %s/<category>/info" % ADDR
+
+
+class CategoryNotFound(DkbApiException):
+    """ Exception indicating that category not found. """
```
Comment (Evildoor) [inline Utils/API/server/lib/dkb/api/exceptions.py]: This was probably meant to be "... that category IS not found.".

**Comment by mgolosova** [comment]
[API-сервер (скелет)](https://trello.com/c/jDycQwXr/5-api-%D1%81%D0%B5%D1%80%D0%B2%D0%B5%D1%80-%D1%81%D0%BA%D0%B5%D0%BB%D0%B5%D1%82)

**Comment by mgolosova** [comment]
Saved original version to [api-orig](https://github.com/PanDAWMS/dkb/tree/api-orig) and rewrote this branch. Main changes:
* `/task/chain`-related stuff is removed (to be moved to another PR);
* `storage`-related stuff is removed (to be moved to another PR);
* `setup.py`: minor refactoring.

Some of the removed stuff will reappear with the first method implementation.

File: Utils/API/server/lib/dkb/api/exceptions.py (hunk 1 above)
Comment (mgolosova) [reply Utils/API/server/lib/dkb/api/exceptions.py:47]: Fixed: #370.

**Comment by mgolosova** [comment]
[Расширение `data4es` для данных по прогрессу обработки событий](https://trello.com/c/Hk5rXhfc/178-%D1%80%D0%B0%D1%81%D1%88%D0%B8%D1%80%D0%B5%D0%BD%D0%B8%D0%B5-data4es-%D0%B4%D0%BB%D1%8F-%D0%B4%D0%B0%D0%BD%D0%BD%D1%8B%D1%85-%D0%BF%D0%BE-%D0%BF%D1%80%D0%BE%D0%B3%D1%80%D0%B5%D1%81%D1%81%D1%83-%D0%BE%D0%B1%D1%80%D0%B0%D0%B1%D0%BE%D1%82%D0%BA%D0%B8-%D1%81%D0%BE%D0%B1%D1%8B%D1%82%D0%B8%D0%B9)

File: Utils/API/server/lib/dkb/api/misc.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
+""" Miscellaneous functions. """
+
+from collections import defaultdict
+
+
+def sort_by_prefixes(values, prefixes, default=0):
+    """ Classify values by 1-char prefixes.
+
+    Generate hash with prefixes as a keys and lists of values -- as value:
```
Comment (Evildoor) [inline Utils/API/server/lib/dkb/api/misc.py:9]:
```suggestion
    Generate hash with prefixes as keys and lists of values -- as values:
```

File: Utils/API/server/lib/dkb/api/misc.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    ```
+    { prefix: [val1, val2, ...], ... }
+    ```
+    If there's no values for some prefix, empty list is set as its value.
```
Comment (Evildoor) [inline Utils/API/server/lib/dkb/api/misc.py:13]:
```suggestion
    If there's no value for some prefix, empty list is set as its value.
```

File: Utils/API/server/lib/dkb/api/misc.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+    :raise TypeError:
+
+    :param values: values to be sorted
+    :type values: list
+    :param prefixes: allowed 1-char prefixes
+    :type profixes: list
+    :param default: id of the prefix to be used ias default when none
```
Comment (Evildoor) [inline Utils/API/server/lib/dkb/api/misc.py:21]:
```suggestion
    :param default: id of the prefix to be used as default when none
```

File: Utils/API/server/lib/dkb/api/storages/__init__.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    """ Get data for histogram of task distribution over time.
+
+    :return: hash of data for the histogram
+    :rtype: dict
+    """
+    return es.task_steps_hist(**kwargs)
+
+
+def task_chain(**kwargs):
+    """ Reconstruct task chain form given task ID.
```
Comment (Evildoor) [inline Utils/API/server/lib/dkb/api/storages/__init__.py:18]:
```suggestion
    """ Reconstruct task chain from given task ID.
```

File: Utils/API/server/lib/dkb/api/storages/exceptions.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    def __init__(self, storage, reason=None):
+        message = "Failed to initialize storage client for '%s'." % storage
+        if reason:
+            message += " Reason: %s" % reason
+        self.details = message
+        super(StorageClientException, self).__init__(message)
+
+
+class NoDataFound(StorageException):
+    """ Exception indicating that requested data not found. """
```
Comment (Evildoor) [inline Utils/API/server/lib/dkb/api/storages/exceptions.py:27]:
```suggestion
    """ Exception indicating that requested data was not found. """
```

File: Utils/API/server/lib/dkb/api/storages/exceptions.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+    def __init__(self, storage, message=None):
+        reason = "No data found in %s" % storage
+        if message:
+            reason += ": %s" % message
+        self.details = reason
+
+
+class QueryException(StorageException):
+    """ Exception indicating any failure in query prepatation. """
```
Comment (Evildoor) [inline Utils/API/server/lib/dkb/api/storages/exceptions.py:38]:
```suggestion
    """ Exception indicating any failure in query preparation. """
```

File: Utils/API/server/lib/dkb/api/storages/es/methods.py [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+    result = transform.task_steps_hist(data, start, end)
+    return result
+
+
+def task_chain(**kwargs):
+    """ Reconstruct task chain from given task ID.
+
+    If task not found in the ES, raises ``NoDataFound``.
+
+    If for given task no ``chain_id`` found, task ID is used instead.
```
Comment (Evildoor) [inline Utils/API/server/lib/dkb/api/storages/es/methods.py:62]:
```suggestion
    If no ``chain_id`` found for given task, task ID is used instead.
```

File: Utils/API/server/lib/dkb/api/storages/es/methods.py [hunk 9]
```diff
@@ -0,0 +1,415 @@
...
+    :param step_type: step definition type: 'step', 'ctag_format'
+                      (default: 'step')
+    :type step_type: str
+
+    :return: hash with calculated statistics for ``step/stat`` method
+             (see :py:func:`api.handlers.step_stat`)
+    :rtype: hash
+    """
+    init()
+    # Aborted/failed/broken/obsolete tasks should be excluded form statistics
```
Comment (Evildoor) [inline Utils/API/server/lib/dkb/api/storages/es/methods.py:370]:
```suggestion
    # Aborted/failed/broken/obsolete tasks should be excluded from statistics
```

File: Utils/API/server/lib/dkb/api/storages/es/transform.py [hunk 10]
```diff
@@ -0,0 +1,575 @@
...
+import logging
+import json
+from datetime import datetime
+
+from api.common import STEP_TYPES
+from common import (ES_DATE_FORMAT,
+                    PREFIX_AGGS)
+
+
+# Common functions for ES respomse parsing
```
Comment (Evildoor) [inline Utils/API/server/lib/dkb/api/storages/es/transform.py:14]:
```suggestion
# Common functions for ES response parsing
```

File: Utils/API/server/lib/dkb/api/storages/es/transform.py [hunk 11]
```diff
@@ -0,0 +1,575 @@
...
+# --->
+
+def steps_iterator(data):
+    """ Gerenator for iterator over steps data.
```
Comment (Evildoor) [inline Utils/API/server/lib/dkb/api/storages/es/transform.py:18]:
```suggestion
    """ Generator for iterator over steps data.
```

File: Utils/API/server/lib/dkb/api/misc.py (hunk 3 above)
Comment (Evildoor) [reply Utils/API/server/lib/dkb/api/misc.py:13]: Or "... there ARE no values ...".

**Comment by mgolosova** [comment]
@Evildoor, typos etc. are fixed: #372.

File: Utils/API/server/lib/dkb/api/storages/es/methods.py [hunk 12]
```diff
@@ -0,0 +1,415 @@
...
+    result = transform.task_steps_hist(data, start, end)
+    return result
+
+
+def task_chain(**kwargs):
+    """ Reconstruct task chain from given task ID.
+
+    If task not found in the ES, raises ``NoDataFound``.
+
+    If no ``chain_id`` was found for given task, its taskiID is used instead.
```
Comment (Evildoor) [inline Utils/API/server/lib/dkb/api/storages/es/methods.py:62]: A suggested change was made here, as well as some other improvements, but now there is a "taskiID".

Comment (mgolosova) [reply Utils/API/server/lib/dkb/api/storages/es/methods.py]: Right, sorry for being inobservant. Fixed now: #373.

**Pull request merged**